### ISSUE TITLE
Intunewin file import issue using import folder settings.

### ISSUE THREAD

**Issue by dominiquestabile** [Micke-K/IntuneManagement#94]
Hello,
Great tool, many thanks for it :)

Intunewin file fails to import doing "Bulk Import" and using the "App package folder" setting after the first file being uploaded.
Sometime two files can be uploaded consecutively.
The console shows the following errors:
- Failed to upload file chunk. The remote server returned an error: (403) Forbidden.
- Failed to finilize upload. The remote server returned an error: (400) Bad Request.
- ![image](https://user-images.githubusercontent.com/43202821/196126395-b9365bf6-3e61-4d26-a1de-2f4eb3c2bbbd.png)

Importing each file individually directedly from the "Applications" tab works almost all the time.
I have the feeling that there is somehow uploads limits.
Doing each app individually and waiting few second between each of them seems to work fine.
And when doing bulk import, after the first fail, every other intunewin file upload will fail too.

Regards.
Dominique

**Comment by Micke-K** [comment]
Hello Dominique,

It does sound like a quota limit but weird that it hasn't been reported before. 

I quickly checked the error which suggests it is a permission problem but it can't be that since it works when you import them one by one. 

I'll see if I can reproduce the error so I can troubleshoot. 

Do you have very large apps that take long time to upload?

Cheers!

**Comment by dominiquestabile** [comment]
Hello Mikael,

Thank you for your reply 😊

Most of my apps are less than 50kb, two are between 25mb and 35mb and two big ones of 650mb.
The issues can happen on any of them independently of the size.
It seems that the first one of each try go through but as soon as one app fails to upload every other one will fail too.
It’s quite easy for me to reproduce, let me know if I can help on something.

Besides this I notices 2 other issues that were not present some weeks ago (with latest and previous version) and a last one since I discovered your tool.

  *   Filters fails to import (working few weeks ago)
  *   Conditional access fails to import (working few weeks ago)
  *   Endpoint security / disk encryption policies import is inconsistent (policy created, but no settings and assignment are broken)

Would you like that I open new issues for each of them ?

BR

Dominique

**Comment by Micke-K** [comment]
I found some issues. Microsoft has change something in Filters and I think last release broke Endpoint Security import
I could not reproduce the Conditional Access import but that could be because you are using a settings that I didn't. If possible, can you attach the json of the CA policy.

I have not had time to look at the App Upload yet.

Extract the EndpointManager.psm1 file and copy/replace the existing on in the Extensions folder. Let me know how any testing goes.

Cheers!

[EndpointManager.zip](https://github.com/Micke-K/IntuneManagement/files/9827923/EndpointManager.zip)

**Comment by Micke-K** [comment]
Try this file. It has two psm1 files. Copy them to the Extensions folder.

Cheers!

[EndpointManager.zip](https://github.com/Micke-K/IntuneManagement/files/9828181/EndpointManager.zip)

**Comment by Micke-K** [comment]
Another version. I tested both Administrative Templates and Endpoint Security policies with this. 

Cheers!

[EndpointManager.zip](https://github.com/Micke-K/IntuneManagement/files/9833650/EndpointManager.zip)

**Comment by dominiquestabile** [comment]
Hi Mickael,
Sorry to answer only now, I had some emergencies to manage.
I just tried the bulk import with Filters and Endpoint Security Policies, it works perfectly now 😊
I saw you already released it, many thanks for the quick reaction.

Let me know for Applications upload.

Have a very nice day.

Dominique

**Comment by Micke-K** [comment]
Hello Dominique,

Thank you for the update. 

I'm building a new test environment and I plan to add 10-15 apps to it and then also build some kind if testing feature into the tool. I need a better way to test export/import to detect bugs and Microsoft changes. This will take some time though as I have lots of other things going on as well. 

Cheers!

**Comment by dominiquestabile** [comment]
Hello Mickael,

Thank you for your reply.
Time is always the biggest issue 😃
Looking forward to test the next release.

BR

Dominique

**Comment by Micke-K** [comment]
Hello Dominique,

Can you try this? Unzip and replace existing file under Extensions.

I could reproduce the error but this code fixed it. I added a retry for 10 s (up to 6 times) and it looks like that works. It always worked after the initial 10s wait. Not happy about the solution since I have not clue why it happens. It's just like the generated SAS token is in the future so it doesn't have access yet. I guess I could see if I could inspect the SAS string and get the time from it but that is for another day.

I have NOT tested installing the app to verify that it works so if you could verify that it would be great!

Cheers!
 
[IntuneAppManagement.zip](https://github.com/Micke-K/IntuneManagement/files/9901359/IntuneAppManagement.zip)

**Comment by dominiquestabile** [comment]
Hello Mickael,
Just tested now, works perfectly 😊
Many thanks

Have a nice day.

Dominique

**Comment by Micke-K** [comment]
Great news! Thank you so much for helping testing it.

Do you still have an issue with the Conditional Access policies?

Cheers

**Comment by dominiquestabile** [comment]
No worries, it’s a win-win 😊
I’m about to test a full bulk import, I’ll let you know if anything fails.
BR

Dominique

**Comment by dominiquestabile** [comment]
Hello Mickael,
Here are the results of my last tests. Obviously the most important is the import but I added the errors I encountered during Bulk Delete and Bulk Export as info.

My import settings:
***@***.***

Issues :


  *   General errors :
     *   Error on Get profile picture (I do not use this, probably an error because it’s empty)
        *   Failed to invoke MS Graph with URL https://graph.microsoft.com/beta/me/photos/48x48/$value (Request ID: 7f7cf0e4-97f6-432f-921c-c7ed9eb8da13). Status code: NotFound

  *   Bulk Import
     *   Error on MS Edge app – Assignements (It happens every time).
        *   In source I have “Required – All devices” and “Available for enrolled devices – All users”
        *   In destination: no assignments
           *   Import Applications object Microsoft Edge for Windows 10 and later
           *   ### TEST ### OData.Type: #microsoft.graph.windowsMicrosoftEdgeApp
           *   Applications object imported successfully with id: 2d875997-1e0f-4eac-9054-fc7169164a19
           *   Failed to invoke MS Graph with URL https://graph.microsoft.com/beta/deviceAppManagement/mobileApps/2d875997-1e0f-4eac-9054-fc7169164a19/assign (Request ID: 2637a713-c4c3-48a0-99d3-cb114219fec9). Status code: BadRequest

     *   Error on Import Enrollment Status Page object All users and all devices (This is a default object, it already exists so that’s probably why, settings ends up the same in source and destination)
        *   Failed to invoke MS Graph with URL https://graph.microsoft.com/beta/deviceManagement/deviceEnrollmentConfigurations//assign<https://graph.microsoft.com/beta/deviceManagement/deviceEnrollmentConfigurations/assign> (Request ID: 00fc4205-1b88-4b37-8e8d-504cfda4ed11). Status code: BadRequest

     *   Error on Import Enrollment Restrictions object All users and all devices (This is a default object, it already exists so that’s probably why, settings ends up the same in source and destination)).
        *   Failed to invoke MS Graph with URL https://graph.microsoft.com/beta/deviceManagement/deviceEnrollmentConfigurations//assign<https://graph.microsoft.com/beta/deviceManagement/deviceEnrollmentConfigurations/assign> (Request ID: ccfb75e1-6e41-411c-8648-4806e132f2bc). Status code: BadRequest
     *   Error on Import Enrollment Restrictions object All users and all devices (The error is shown twice, This is a default object, it already exists so that’s probably why, settings ends up the same in source and destination)).
        *   Failed to invoke MS Graph with URL https://graph.microsoft.com/beta/deviceManagement/deviceEnrollmentConfigurations//assign<https://graph.microsoft.com/beta/deviceManagement/deviceEnrollmentConfigurations/assign> (Request ID: 5ec77905-5af7-4689-8c28-d43ee8c6c244). Status code: BadRequest

     *   Error on Import Compliance Policies object Standardkonformitätsrichtlinie für Android (This is a default object, it already exists so that’s probably why, settings ends up the same in source and destination)).
        *   Failed to invoke MS Graph with URL https://graph.microsoft.com/beta/deviceManagement/deviceCompliancePolicies (Request ID: 751d00bf-cc72-4532-b3f8-bd68c515cbf0). Status code: BadRequest

     *   Error on Import Conditional Access object MFA (Nothing is imported – Error from source - Fixed)
Microsoft Intune Enrollment app is present in source but does not exist anymore, removing it from source did solve the issue
        *   Failed to invoke MS Graph with URL https://graph.microsoft.com/beta/identity/conditionalAccess/policies (Request ID: b0437fbd-00b1-48d2-8d50-9770cef98d2b). Status code: BadRequest



  *   Bulk Delete ( not important but for info)
     *   Error on Get App Protection objects, (I do not use this, probably an error because it’s empty)
        *   Delete  object
        *   Failed to invoke MS Graph with URL https://graph.microsoft.com/beta/ (Request ID: e6c2eeae-dd50-4608-9138-566eeba960a2). Status code: MethodNotAllowed
     *   Error on Delete Applications object Company Portal (this is synced from MS Store for Business, that’s probably why).
        *   Failed to invoke MS Graph with URL https://graph.microsoft.com/beta/deviceAppManagement/mobileApps/6bd3b98b-280f-484b-99fd-c1dcdabb6b50 (Request ID: 30892f61-bbc3-4031-9de4-2a44e65552e4). Status code: BadRequest
     *   Error on Get Endpoint Security objects (Occured only once)
        *   Delete  object
        *   Failed to invoke MS Graph with URL https://graph.microsoft.com/beta/ (Request ID: 2913eadb-e7e3-4d23-a7b9-05fe86fcf451). Status code: MethodNotAllowed
     *   Error deleting one of the filters (Occured only once)
        *   Delete Filters object Filter_AVDHosts
        *   Delete Filters object Filter_AVDHosts
        *   Failed to invoke MS Graph with URL https://graph.microsoft.com/beta/deviceManagement/assignmentFilters/9c8acec9-d775-43d1-8720-9b099ebc301e (Request ID: d85b7034-4338-4434-872d-3722dd373194). Status code: BadRequest
        *   Delete Filters object Filter_Windows10
        *   Delete Filters object Filter_Windows11
     *   Error on Policy Sets objects (we do not use this, probably an error because it’s empty)
        *   Get Policy Sets objects
        *   Delete  object
        *   Failed to invoke MS Graph with URL https://graph.microsoft.com/beta/ (Request ID: aa1aa0f7-8dd1-4c30-875f-e34ccdbbcc5c). Status code: MethodNotAllowed
  *   Bulk Export ( not important but for info)
     *   Error on Export App Protection (I do not use this and message is pretty obvious)
        *   No object to export

     *   Error on Export Azure Branding (I do not use this and message is pretty obvious)
        *   No object to export

     *   Error on Export Policy Sets (we do not use this, probably an error because it’s empty)
        *   Failed to invoke MS Graph with URL https://graph.microsoft.com/beta/deviceAppManagement/policySets/?$expand=Items (Request ID: 9d96be09-c3fe-45ac-8219-b2eee0091d31). Status code: BadRequest
        *   No object to export

     *   Error on roles export, our custom roles assignments use “Scope tags” and “Members” is not populated.
        *   Export Role Definitions: IntuneRole_Apps_Management
        *   Failed to invoke MS Graph with URL https://graph.microsoft.com/beta/groups/1b8949a5-a161-49f3-ad4d-60c842863b38 (Request ID: da428aa8-46b1-4562-b41b-a0e2db94b8fc). Status code: NotFound
        *   Failed to invoke MS Graph with URL https://graph.microsoft.com/beta/groups/8192ed5c-6772-4fc9-bbae-ff22670cdb06 (Request ID: b9a6a23d-db78-40c9-a702-f811e03d58f8). Status code: NotFound
        *   Export Role Definitions: IntuneRole_Devices_Management
        *   Failed to invoke MS Graph with URL https://graph.microsoft.com/beta/groups/6772957f-e853-4c54-96df-214b5d825494 (Request ID: 13dc0305-ceee-4d85-8a99-bbef7f0edbbd). Status code: NotFound
        *   Failed to invoke MS Graph with URL https://graph.microsoft.com/beta/groups/fafb5635-b528-4444-8645-0eaae96890f2 (Request ID: a9dd8915-f722-4526-a357-37087a86166d). Status code: NotFound
        *   Failed to invoke MS Graph with URL https://graph.microsoft.com/beta/groups/e90055ed-c6c4-4304-a236-81875d96e831 (Request ID: 0186b4c5-0e89-42bf-a83a-d1450a91c5cd). Status code: NotFound
        *   Failed to invoke MS Graph with URL https://graph.microsoft.com/beta/groups/8192ed5c-6772-4fc9-bbae-ff22670cdb06 (Request ID: ca884717-9e33-4f6b-b5a0-3b23dd4c8062). Status code: NotFound

     *   Error on Export Terms of use objects
        *   ----------------------------------------------------------------
        *   Failed to invoke MS Graph with URL https://graph.microsoft.com/beta/identityGovernance/termsOfUse/agreements?$expand=files (Request ID: f6a4cad9-776a-4cfe-ac55-98b5a935b2bd). Status code: Forbidden

Let me know if I can test something else and thanks again for everything.

BR

Dominique

**Comment by Micke-K** [comment]
Hello!

Thank you so much for that list. I think it looks worse than it is. I'll see if I can start looking at them this or next week.

I am planning to implement a new tool in the script. The plan is to do all this in an automated way eg. Import everything from a specific folder, check that assignments are added, then export all of them, copy each imported object and then delete all of them. I have the plan in my head but not enough time at the moment so I really appreciate all the help with testing.    

Terms of use objects - Do you have all the permissions for this and have you tested with the latest version I uploaded to Issue 27 the other day? This file: [EndpointManager.zip](https://github.com/Micke-K/IntuneManagement/files/9877282/EndpointManager.zip). It now supports exporting the PDF file.

Cheers!

**Comment by dominiquestabile** [comment]
Terms of use pdf import and export works fine:
***@***.***

BR
Dominique

**Comment by Micke-K** [comment]
Hello!
I THINK this will take care most of the issues. They were not really any major issues but mostly bad handling of empty arrays, reporting, and other minor issues.
I have not looked at the "Compliance Policies" issue yet. I'll see if I can replicate that next week.

- MS Edge app assignment: Bug fix. I had to remove two additional properties before import.
- Assignment for default policies (Enrolment Restrictions, ESP etc.)
- Bulk Delete: Band handling for types with no policies, order bulk delete (filters cannot be removed if they are used), Business Store Apps will not be "deleted"
- Removed error for profiles without photos
- Removed errors for getting groups that are not found. Looks like your Custom Role assignment once had groups assigned to them but the script could not find the groups during import. At least that is my understanding of it.

I also added additional info when it fails to call a Graph API. This will now add Response Message eg instead of just saying Bad Request, it will add that a filter cannot be deleted because it is assigned.

Probably a few other minor things I've fixed that I cannot think of now. Let me know if you test it and how it goes...

Cheers!

[Extensions.zip](https://github.com/Micke-K/IntuneManagement/files/9980915/Extensions.zip)

**Comment by dominiquestabile** [comment]
Hello Mikael,
Thank you very much.
I’m working on a big issue on the moment, I will be able to test that on Monday and I will let you know.
I forgot to mention you another issue with Apps. The dependencies are not imported.
Thanks again for everything, it’s great job that you’re doing.

BR

Dominique

**Comment by Micke-K** [comment]
Hello!

Dependencies should work. Do you have any examples/logs when it is failing?

Thank you for all the help. 

Cheers!

**Comment by Micke-K** [comment]
Hello

Just checking in to see if there are any outstanding issues on this or if I can close the issue?

Cheers!

**Comment by dominiquestabile** [comment]
Hi Mikael,
Sorry to come back to you only now.
End/beginning of the year has been bumpy… A lot of stuff to finish, holidays, etc…

In the last imports we did everything seemed to work as expected.
All fine here.
We used your solution, on a pilot phase with few customers “manually”.
We are now working on the automation side and there is big probability that we use your tool, at least at the beginning.
I’ll keep you posted if anything new comes up.
Thanks a lot for your help.

BR

Dominique
